### PR TITLE
[bridge-indexer] Progress saving policy

### PR DESCRIPTION
**Pull request by longbowlu**
## Description 

This PR introduces `ProgressSavingPolicy` to deal with two problems:
1. The current implementation has a bug on Sui side - checkpoints data arrive out-of-order (e.g. checkpoint 10 may be processed earlier than checkpoint 9), so existing `save_process` may cause us to miss blocks.
2. In current implementation we need to write progress to DB for every call to `save_process`. This can be optimized to cache progresses in memory and flush them periodically or conditionally. 

We add two types of `ProgressSavingPolicy`, `SaveAfterDuration` and `OutOfOrderSaveAfterDuration`:
* `SaveAfterDuration` only flushes the progress to DB after a period of time
* `OutOfOrderSaveAfterDuration` assumes the data is out of order, and will only write height N when it makes sure everything before N has been received.
* 

## Test plan 

unit tests and production deployment.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Review by dariorussi — APPROVED** [review]
It looks good to me.
Do you think we need all the customization provided?
I feel an in memory story with all DB updates done with some interval (interval in checkpoints and not time) would be best.
Maybe that is what we do, but I am getting a bit lost in the level of abstraction provided by our indexer.

One question: is this affecting the deepbook indexer?

**Comment by longbowlu** [comment]
> It looks good to me. Do you think we need all the customization provided? I feel an in memory story with all DB updates done with some interval (interval in checkpoints and not time) would be best. Maybe that is what we do, but I am getting a bit lost in the level of abstraction provided by our indexer.
> 
> One question: is this affecting the deepbook indexer?

These two basic policies are what we need for Sui ingestion and Eth ingestion:
* SaveAfterDuration does what you described in time. I think we can also do in checkpoints but then we need to allow custom value for checkpoint length cuz it may vary based on chains. We will use this for Eth, because eth data always comes in strict order.
* OutOfOrderSaveAfterDuration is made for sui because when the data may come out of order from data ingestion framework. It means we need to re-order the data to make sure nothing is missed.
To summarize these two policies are tailored to what we need today for bridge indexer.

**Comment by longbowlu** [comment]
> One question: is this affecting the deepbook indexer?

yes it does, if DB indexer uses this framework (this is why @0xaslan is in the reviewer list :)
I think DB indexer can just use `SaveAfterDuration` policy.
